### PR TITLE
Optionally sign initial SOA query

### LINKED ISSUE #9671: rfc2136 broken with hidden primary setups
https://github.com/certbot/certbot/pull/9408 has caused certbot to fail in hidden master setups because the SOA query fails due to an unknown TSIG key.

For security reasons, keys are only stored on a hidden primary, hence the secondary rejects the TSIG signed dns query.

suggest this commit be reverted or logic added to make TSIG signatures on DNS queries optional.
Fundamentally, TSIG is only required on DDNS updates (and zone transfers)

### PR DESCRIPTION
**Pull request by frillip**
Initially implemented (but non-configurable) in #9408 , raised as a regression by @benweeks2 in #9671 . This adds an optional configuration parameter to enable signing of the initial SOA query when determining the authoritative nameserver for the zone. Default is disabled.

Documentation has also been updated.

Added @warewolf to `AUTHORS.md` as a lot of this has come from his original PR on #9106 

Fixes #9671.

## Pull Request Checklist

- [x] The Certbot team has recently expressed interest in reviewing a PR for this. If not, this PR may be closed due our limited resources and need to prioritize how we spend them.
- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [x] Include your name in `AUTHORS.md` if you like.

### CODE REVIEW

File: certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ class Authenticator(dns_common.DNSAuthenticator):
     }
 
     PORT = 53
+    sign_query = "False"
```
Comment (alexzorin) [inline certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py:43]: Please follow the existing pattern for defining credential options, have a look at `_setup_credentials` and `_validate_credentials`.

**Review by alexzorin — CHANGES_REQUESTED** [review]
My deep thanks for hopping on this so quickly and also for being responsive about the regression. The RFC2136 plugin is tricky to balance and your input is really helpful.

I've left some feedback and you will also need to fix the test suite.  https://eff-certbot.readthedocs.io/en/stable/contributing.html#running-a-local-copy-of-the-client goes into some detail about getting a development environment up and running and how to run the tests.

File: certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py [hunk 2]
```diff
@@ -94,21 +95,28 @@ def _get_rfc2136_client(self) -> "_RFC2136Client":
...
         self.port = port
         self.keyring = dns.tsigkeyring.from_text({
             key_name: key_secret
         })
         self.algorithm = key_algorithm
+        if sign_query.upper() == "TRUE":
+            self.sign_query = True
+        else:
+            self.sign_query = False
+
```
[the remark this answers is not in the document]
Comment (frillip) [reply certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py:119]: I did initially try this as a bool, but realised that `self.credentials.conf('sign_query')` was pulling it in as a string when `= false` in the ini file gave me `True`, so opted to pass the string as is out the config file, or the default value I set earlier on in `self.sign_query`, so it should never have been `None`. But this way makes much more sense, thanks.

**Comment by frillip** [comment]
I can imagine it's tricky to balance! I've worked with some hidden primary setups before, but in those instances we always sent updates to the primary direct, rather than via the secondaries. 

I'll try and find some time to fix the test suite failures this evening.

File: certbot/CHANGELOG.md [hunk 3]
```diff
@@ -10,7 +10,8 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
 
 ### Changed
 
-*
+* Optionally sign the SOA query for dns-rfc2136, to help resolve problems with split-view
```
[the remark this answers is not in the document]
Comment (frillip) [reply certbot/CHANGELOG.md:13]: Added some more detail here.

File: certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py [hunk 4]
```diff
@@ -89,26 +90,34 @@ def _cleanup(self, _domain: str, validation_name: str, validation: str) -> None:
...
+        if cast(str, self.credentials.conf('sign_query')).upper() == "TRUE":
+            sign_query = True
+
+        return _RFC2136Client(cast(str, self.credentials.conf('server')),
+                              int(cast(str, self.credentials.conf('port')) or self.PORT),
+                              cast(str, self.credentials.conf('name')),
+                              cast(str, self.credentials.conf('secret')),
+                              self.ALGORITHMS.get(self.credentials.conf('algorithm') or '',
+                                                  dns.tsig.HMAC_MD5),
+                              sign_query)
```
Comment (alexzorin) [inline certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py:104]: This still crashes Certbot if `dns_rfc2136_sign_query` is absent from the credentials file. You can just replace `sign_query` with:

```python
(self.credentials.conf('sign_query') or '').upper() == "TRUE"
```

File: certbot/CHANGELOG.md [hunk 5]
```diff
@@ -10,7 +10,16 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
...
+  DNS setups and hidden primary setups.
+  * Certbot versions prior to v1.32.0 did not sign queries with the specified TSIG key
+    resulting in difficulty with split-horizon implementations.
+  * Certbot versions later than v1.32.0 signed queries by default, potentially causing
```
Comment (alexzorin) [inline certbot/CHANGELOG.md:17]:
```suggestion
  * Certbot v1.32.0 through v2.5.0 signed queries by default, potentially causing
```

File: certbot/CHANGELOG.md [hunk 6]
```diff
@@ -10,7 +10,16 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
...
+    incompatibility with hidden primary setups with `allow-update-forwarding` enabled
+    if the secondary did not also have the TSIG key within its config.
+  * Certbot versions later than v2.6.0 now do not sign queries by default, but allow
```
Comment (alexzorin) [inline certbot/CHANGELOG.md:20]:
```suggestion
  * Certbot v2.6.0 and later no longer signs queries by default, but allows
```

**Review by alexzorin — CHANGES_REQUESTED** [review]
Thanks for updating this. Apart from the feedback in comments, please also merge `master` and resolve the two conflicts in the CHANGELOG and `dns_rfc2136.py`.

File: certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py (hunk 4 above)
Comment (frillip) [reply certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py:104]: I did test this with the statement removed from the credentials file and worked with no crashes on my end, but again, this seems to be a more elagant way of doing this. Thanks!

**Review by alexzorin — APPROVED** [review]
LGTM, thanks.

**Pull request merged**